### ISSUE TITLE
prepare data for KNB

### ISSUE THREAD

**Issue by cboettig** [boettiger-lab/anna-eml#13]
Hi @liuanna ,

Okay, got feedback from @mbjones on the GPDD example and he agrees it would be worth providing the atomized EML files and associated CSVs to the KNB database.  Recall that the atomized EML files contain a lot of information (particularly coverage) specific to individual datasets and not represented in the overall EML file for the database as a whole.  Consequently current searches of KNB for a species or geographic range covered by the GPDD will not show up any GPDD data, since the search system only indexes what is in the EML.  

I think the simplest way to go about this is to add a directory to this repo, containing all the atomized EML files and their associated CSV tables, for all datasets in the GPDD (i.e. by just running your script over the whole dataset).  Then we can touch base again with Matt about how best to upload it to the existing KNB entry for the GPDD.

**Comment by liuanna** [comment]
Sounds good. I am doing this task now. But it seems take too long to generate total 4471 eml files corresponding to each main ID. In current speed, it will take over 30 hours.

**Comment by cboettig** [comment]
Yup, it's a lot of files.  I can run the script on our server then so it doesn't tie up your machine for that long

**Comment by liuanna** [comment]
Will I be able to oversee it in the process? Since the script may break somewhere I don't know. I wrote the script based on the first 10 main ID and the general pattern. However, the whole dataset contains some special cases. For example, some main IDs do not have data and time period files. Some files do not contain location coordinates (just NA), and our ```set_coverage``` function does not accept NA. I have fixed these two problems in my script, but I don't know if other problems will arise in the process.

**Comment by cboettig** [comment]
Good point, I'll just email you the login and you can oversee the Ron from the rstudio server

**Comment by liuanna** [comment]
Great. Thank you!

**Comment by liuanna** [comment]
I have set up, and the script is running on the server now. I will start working on other tasks and come back to this one as long as the script finishes generating files.

**Comment by liuanna** [comment]
Hi @cboettig, I finally finish generating all files. I create a directory called `file` and create a folder for each main ID containing an eml file and all associated csv files. However, there are too many files. GitHub said "we had to truncate this directory to 1,000 files. 4,156 entries were omitted from the list".

**Comment by cboettig** [comment]
Hi Anna,

Great work!  No worries, that message is just in regards to the web display; all the files are still on GitHub; and I see all 5156 when I pull the repo.  Have you had a chance to run the validator on all the EML files? That would be a good next step while the server instance is still up (will probably take a while).

**Comment by liuanna** [comment]
@cboettig Great! I have validated all eml files. It took about two days to run and validate all files and finally finished this afternoon.

**Comment by cboettig** [comment]
This looks great.  Just waiting to hear from @mbjones about how best to proceed with uploading these to KNB.

**Comment by mbjones** [comment]
The EML files look great.  I think we need to think about our use cases in trying to decide how to structure this in the KNB.  Dividing the GPDD up into lots of data sets provides a nice exposure of the EML that you created, but means we would have thousands of separate data sets, which is much harder to manage.

## Option: Nested data packages
We could nest them hierarchically in a parent 'GPDD' package, with children for specific clusters of the data.

The [Levy dataset](https://knb.ecoinformatics.org/#view/doi:10.5063/F1Z899CZ) is an example of a hierarchical data set that has 19 child data sets and divides up its content into logical tiers and files.  The advantage is that smallish chunks of data can be downloaded if that is all someone wants (the whole data set is 2TB). The disadvantage is that you have to download the pieces separately. The GPDD data set would have 5157 tiny children (I think), which may be too unwieldy to be useful.

One possibility would be to come up with multiple levels of package hierarchy (such as taxon, time, or space).  So, for example, we might have a package hierarchy like the following:

```
GPDD
|---- Mammals
        |---- 1900-1950
        |---- 1951-2000
        |---- 2001-2016
|---- Insects
        |---- 1900-1950
        |---- 1951-2000
        |---- 2001-2016
|---- Birds
        |---- 1900-1950
        |---- 1951-2000
        |---- 2001-2016
```

I'm not sure what the right organization is, but a browse hierarchy that provided some organization to the data sets might help.  

One of the things we did for the Levy data set was to provide a [custom R function](https://knb.ecoinformatics.org/#view/urn:uuid:1c865ce5-5fe9-4b49-a123-f755c26261f3) that took some selection parameters and automatically located the right subset of data and downloaded it.  This helps users to reassemble the massively subsetted data into aggregates that are useful. But people have to know how to use R for that.

## Option: extended metadata
Ultimately, the GPDD is a pretty small data set, about 10MB. So, downloading it all at once is easy and it really doesn't make sense to divide it into literally thousands of files.  An option would be to keep it as a single database, but massively supplement the metadata record by putting in all of the individual geographic, temporal, and taxonomic coverages into the current main EML record.  This would make the taxa, temporal coverage, and spatial coverage fields all available for searching, but ultimately the same GPDD record would come up on any match.

Let's discuss please.

**Comment by cboettig** [comment]
Thanks Matt, this is great, exactly the issues I was thinking about too.  I think it would be good to document these considerations and what we end up going with, as they speak to more general issues.  In some sense these options highlight the challenge in balancing usability with extensibility; and perhaps the related trade-off between making it easier to use data vs making it easier to contribute data.

I suspect the most common use case would still be using the GPDD as a whole, and like you say we don't want to make this extremely difficult / requiring a lot of specialize R code.  Given the small file size, Option 2: extended metadata clearly does this best.

Still, I think this leaves open some questions about extensibility & generality.  First, I completely agree with you about the file size being relevant to the design strategy.  With rOpenSci we've wrapped a lot of APIs now that have some clever query design so that you only download the subset of the data you need, but 9 times out of 10 this is just a nuisance and you end up looping over API calls when you would rather just download all the data and subset it locally.  (After all, I suppose the APIs are often supporting a very different use case where the goal is to display a small amount of data on a webpage as fast as possible, minimizing download & client-side processing; while for scientific work we're happy to wait a few seconds to download locally and happy to do most subsetting / processing on the client). So if the only real advantage to option 1 is the ability to do more server-side filtering/sub-setting and thus provide smaller downloads, the added complexity probably isn't worth it. (Though that script in the Levy data does show a compelling use case for the dataone R package!)

To some extent, the non-nested option may also make it harder to re-use / recombine elements of the GPDD in other analyses; though I think the extended metadata goes a long way to addressing this. 

Option 2 also might not be best for making it easy to extend the GPDD. There is at least some interest in this, e.g. see this thread: https://github.com/ropensci/rgpdd/issues/5

In a completely idealized version of a metadata driven repository, a dataset like the GPDD could be assembled on the fly from individual studies appearing as completely separate units on KNB.  Such a synthetic dataset would be extended every time someone uploads a new population dynamics data set to KNB.  Of course that won't work without some magical semantics most contributors wouldn't use anyway. But if all of those contributions followed a rough schema / template and maybe included an appropriate tag or other identifier, it might be possible with less manual effort than it took to build the original GPDD at any rate.  This synthesis could be deposited as a separate, derived dataset in KNB, ideally with scripts to re-generate it.  Perhaps this could be approximated by doing both options, or having a nested structure that also includes the synthesized tables, e.g.

```
GPDD
|---- All
|---- Mammals
        |---- 1900-1950
        |---- 1951-2000
        |---- 2001-2016
|---- Insects
        |---- 1900-1950
        |---- 1951-2000
        |---- 2001-2016
|---- Birds
        |---- 1900-1950
        |---- 1951-2000
        |---- 2001-2016
...

```

I'm curious what you think about having both derived/synthesized and raw/atomized versions of the same data on KNB. On one hand it seems fundamentally poor form to be duplicating the same data, but on the other it does seem to be consistent with your diagram in [Reichman et al](http://doi.org/10.1126/science.1197962). 

On a more technical note, how would the EML be structured in the case of the extended metadata?  Just use repeating `<coverage>` elements for each of the 5157 studies in the database?  Where would the information about the authors of individual data sets go?
